text/0080-datetime-extension.md

```diff
@@ -0,0 +1,488 @@
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.now.timestamp.offset(principal.timeZoneOffset).toTime() >= duration("9h") &&
+  context.now.timestamp.offset(principal.timeZoneOffset).toTime() <= duration("17h")
+};
+```
+
+### Should we allow operator overloading?
+
+The current proposal supports allowing overloading the operators `<`, `<=`, `>`, `>=`, and `-`.
+If this is an unacceptable change to the language, then we could instead provide special names for these functions, like the ones listed below.
+Note that names like `.lessThan()` are already reserved by the decimal extension, and so cannot be used for `datetime` or `duration`.
+
+- `.isBefore()` check if a date is before another (non-inclusive)
+- `.isAfter()` check if a date is after another (non-inclusive)
+- `.isShorterThan()` check if a duration is shorter than another (non-inclusive)
+- `.isLongerThan()` check if a duration is longer than another (non-inclusive)
+
+`==` and `!=` will work by default since these operations are defined for any type.
+
+### Should we allow negative durations?
+
+The current proposal supports both positive and negative durations.
+A negative duration may be useful when a user wants to use `.offset()` to shift a date backwards.
+For example: `context.now.offset(duration("-3d"))` expresses "three days before the current date".
+
+But some users may find the concept of negative durations confusing, so we have also considered some alternative definition of subtraction (`-`) that will allow us to avoid them:
+
+- Define `DT1.durationSince(DT2)` as the _absolute value_ of the difference between `DT1` and `DT2`.
+- Define `DT1.durationSince(DT2)` to be zero when `DT2 > DT1`.
+- Define `DT1.durationSince(DT2)` to be an error when `DT2 > DT1`.
+
+### Is millisecond precision enough? (Or too much?)
+
+The current proposal supports milliseconds. Do we want nanoseconds too? Or is just seconds sufficient? The ISO 8601 format does not specify a maximum precision, so we can technically allow any number of `S`s after the `.` in `YYYY-MM-DDThh:mm:ss.SSSZ`. Based on [this blog post](https://nickb.dev/blog/iso8601-and-nanosecond-precision-across-languages/), it appears that Javascript supports milliseconds (3 digits), Python supports microseconds (6 digits), and Rust and Go support nanoseconds (9 digits). Assuming nanosecond accuracy, the maximum (signed) 64-bit number (2^63 - 1) represents April 11, 2262. This date seems far enough out that any of these choices (milliseconds, microseconds, or nanoseconds) seems reasonable.
+
+Note that the backwards compatible option is to stick with milliseconds (or seconds). We can add precision later, but not remove it without a breaking change.
```

Comment:
In what sense is it backward compatible? The representation (as a `long`) would change its meaning, which means that data stored in the app would have to change. Are we expecting never to have to store a `long` (for duration or time) in the app? I was thinking "current time" or various deadlines would be stored that way.

Reply:
Current time will need to be passed as a ISO date string with the current proposal. Users should not use the Unix `long` value directly.

Increasing precision is backwards compatible in the sense that all ISO date strings that were previously accepted will continue to be accepted in the future. But you are right that this isn't truly backwards-compatible: more precision will affect the underlying representation, which will impact when certain operators will overflow. My comment was sweeping this under the rug 🙂

Reply:
Note that there isn't a (nice) way to make it backwards compatible for analysis because extending the bitvector representation won't correspond to exactly to N extra decimal digits.